Kernel/Bus/USB/UHCI/UHCIDescriptorTypes.h

```diff
@@ -252,38 +267,64 @@ static_assert(AssertSize<TransferDescriptor, 32>()); // Transfer Descriptor is a
 // Description here please!
 //
 struct alignas(16) QueueHead {
+#if ARCH(X86_64)
+    static constexpr size_t DESCRIPTOR_PAD_BYTES = 16u;
+#elif ARCH(I386)
+    static constexpr size_t DESCRIPTOR_PAD_BYTES = 20u;
+#else
+#    error "QueueHead padding not defined for this platform!"   // Error out for now
+#endif
+    struct QueueHeadBookkeeping {
+#if ARCH(X86_64)
+        using QueueHeadDescriptorPointer = Ptr64<QueueHead>;
+        using TransferDescriptorPointer = Ptr64<TransferDescriptor>;
+        using TransferPointer = Ptr64<Transfer>;
+#elif ARCH(I386)
+        using QueueHeadDescriptorPointer = Ptr32<QueueHead>;
+        using TransferDescriptorPointer = Ptr32<TransferDescriptor>;
+        using TransferPointer = Ptr32<Transfer>;
+#endif
+        u32 paddr { 0 };                                // Physical address where this QueueHead is located
+        QueueHeadDescriptorPointer next_qh { nullptr }; // Next QH
+        QueueHeadDescriptorPointer prev_qh { nullptr }; // Previous QH
+        TransferDescriptorPointer first_td { nullptr }; // Pointer to first TD
+        TransferPointer transfer { nullptr };           // Pointer to transfer linked to this queue head
+        bool in_use { false };                          // Is this QH currently in use?
+    };
+
     enum class LinkPointerBits : u32 {
         Terminate = 1,
         QHSelect = 2,
     };
 
     QueueHead() = delete;
     QueueHead(u32 paddr)
```

Comment:
u32 physical address 🤔 

Is it a constraint of the USB bus that these must always have a paddr in the lowest 4 GiB? Or does this change need more help to be 64-bit friendly

Reply:
> u32 physical address 🤔
> 
> Is it a constraint of the USB bus that these must always have a paddr in the lowest 4 GiB? Or does this change need more help to be 64-bit friendly

Yep. As @b14ckcat has said, this is a restriction of the hardware itself. The address, in actuality, is used in an arithmetic way because the descriptors need to be written with the address of the descriptor that's chained to it. Initially when I first wrote this I used a more concrete type here, but it resulted in a load of `static_cast<>`s for no real benefit because these pointers are _never_ dereferenced (as you can see in the member functions). See here for how this is generated: https://github.com/SerenityOS/serenity/blob/master/Kernel/Bus/USB/UHCI/UHCIDescriptorPool.h#L78

Reply:
> This brings up another good question though, are kernel regions as used by the host controller driver here guaranteed to be backed by physical memory within the first 4GB of memory?

This is a good point, and we may need to refactor this later. We should probably be allocating a DMA buffer page to ensure that this is in a range that's acceptable for the hardware, _or_ create another allocation function that ensures that any allocation is below the 4GiB mark.